packages/versioning/src/versioning.ts

```diff
@@ -0,0 +1,154 @@
+import { NumericLiteralType, Program, StringLiteralType, Type } from "@cadl-lang/compiler";
+const addedOnKey = Symbol();
+const removedOnKey = Symbol();
+const versionsKey = Symbol();
+const renamedFromKey = Symbol();
+const madeOptionalKey = Symbol();
+
+export function $added(p: Program, t: Type, v: StringLiteralType | NumericLiteralType) {
+  p.stateMap(addedOnKey).set(t, v);
+}
+export function $removed(p: Program, t: Type, v: StringLiteralType | NumericLiteralType) {
+  p.stateMap(removedOnKey).set(t, v);
+}
+export function $renamedFrom(
+  p: Program,
+  t: Type,
+  v: StringLiteralType | NumericLiteralType,
+  oldName: StringLiteralType
+) {
+  const record = { v: v, oldName: oldName };
+  p.stateMap(renamedFromKey).set(t, record);
+}
+
+export function $madeOptional(p: Program, t: Type, v: string | boolean) {
+  p.stateMap(madeOptionalKey).set(t, v);
+}
+
+export function getRenamedFromVersion(p: Program, t: Type) {
+  return p.stateMap(renamedFromKey).get(t)?.v ?? -1;
+}
+export function getRenamedFromOldName(p: Program, t: Type) {
+  return p.stateMap(renamedFromKey).get(t)?.oldName ?? "";
+}
+export function getAddedOn(p: Program, t: Type) {
+  return p.stateMap(addedOnKey).get(t) ?? -1;
+}
+export function getRemovedOn(p: Program, t: Type) {
+  return p.stateMap(removedOnKey).get(t) ?? Infinity;
+}
+
+export function getMadeOptionalOn(p: Program, t: Type) {
+  return p.stateMap(madeOptionalKey).get(t) ?? -1;
+}
+
+export function $versioned(p: Program, t: Type, v: Type) {
+  const versions = [];
+
+  switch (v.kind) {
```

Comment:
In this case, what kind of model should 't' be?  I think the reason for this property is to allow custom ordering of versions.  I am not sure we would want to allow this more than once in a document, and so it probably needs to apply to the global namespace.

Also, wondering if we should require this - we might want to collect versions based on application of decorators and apply the implied ordering (numeric, lexicographic) of the version type, if this is not specified.

Reply:
`t` right now has to be a namespace, the intention is you put it along with the rest of your service metadata.

I feel like it's best to be intentional about declaring a new version. Comparing versions seems like sort of a boondoggle. Numeric ordering is relatively easy. Lexicographic seems like a footgun - probably you want semver comparison instead?

Reply:
We are catching references to non-existent versions in the individual version decorators, so I think this comment is moot.